### PR TITLE
Editorial: Add new convention for abstract ops that don't throw

### PR DESCRIPTION
**Pull request by bterlson**
As discussed in #253. Thoughts on this? I personally find this easier to read.

I've also gone ahead and added ! to a few places where it is needed (mostly ToString invocations). The next biggest class of calls that can't fail are to Type and min/max/friends. Since these can never throw by definition I wonder if it's worth adding ! in these types of cases. I'm leaning toward not on the grounds that Type is a short-hand and min/max/etc. are math functions and so neither actually returns a completion record... Thoughts?

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -13851,7 +13848,7 @@
               1. Let _nextValue_ be IteratorValue(_next_).
               1. If _nextValue_ is an abrupt completion, set _iteratorRecord_.[[done]] to *true*.
               1. ReturnIfAbrupt(_nextValue_).
-              1. Let _status_ be CreateDataProperty(_A_, ToString(_n_), _nextValue_).
+              1. Let _status_ be CreateDataProperty(_A_, ! ToString(_n_), _nextValue_).
```
Comment (ljharb) [inline spec.html]: should `CreateDataProperty` have a `?` or a `!` in front of it?

File: spec.html [hunk 2]
```diff
@@ -15139,7 +15135,7 @@
             1. Let _nextValue_ be IteratorValue(_next_).
             1. If _nextValue_ is an abrupt completion, set _iteratorRecord_.[[done]] to *true*.
             1. ReturnIfAbrupt(_nextValue_).
-            1. Let _status_ be CreateDataProperty(_A_, ToString(_n_), _nextValue_).
+            1. Let _status_ be CreateDataProperty(_A_, ! ToString(_n_), _nextValue_).
```
Comment (ljharb) [inline spec.html]: also here

File: spec.html [hunk 3]
```diff
@@ -36719,8 +36709,7 @@
                     1. Let _genvRec_ be _genv_'s EnvironmentRecord.
                     1. Let _benv_ be the running execution context’s LexicalEnvironment.
                     1. Let _benvRec_ be _benv_'s EnvironmentRecord.
-                    1. Let _fobj_ be _benvRec_.GetBindingValue(_F_, *false*).
-                    1. Assert: _fobj_ is never an abrupt completion.
+                    1. Let _fobj_ be ! _benvRec_.GetBindingValue(_F_, *false*).
                     1. Let _status_ be ? _genvRec_.SetMutableBinding(_F_, _fobj_, *false*).
```
Comment (ljharb) [inline spec.html]: this `_status_` is never used, should it be `Perform ? _genvRec_.SetMutableBinding(_F_, _fobj_, *false*).`?

File: spec.html [hunk 4]
```diff
@@ -36640,10 +36633,8 @@
                   1. Let _fenvRec_ be _fenv_'s EnvironmentRecord.
                   1. Let _benv_ be the running execution context's LexicalEnvironment.
                   1. Let _benvRec_ be _benv_'s EnvironmentRecord.
-                  1. Let _fobj_ be _benvRec_.GetBindingValue(_F_, *false*).
-                  1. Assert: _fobj_ is never an abrupt completion.
-                  1. Let _status_ be _fenvRec_.SetMutableBinding(_F_, _fobj_, *false*).
-                  1. Assert: _status_ is never an abrupt completion.
+                  1. Let _fobj_ be ! _benvRec_.GetBindingValue(_F_, *false*).
+                  1. Let _status_ be ! _fenvRec_.SetMutableBinding(_F_, _fobj_, *false*).
```
Comment (ljharb) [inline spec.html]: This `_status_` is never used, should it be `Perform ! _fenvRec_.SetMutableBinding(_F_, _fobj_, *false*).`?

File: spec.html [hunk 5]
```diff
@@ -36669,8 +36660,7 @@
                     1. Let _genvRec_ be _genv_'s EnvironmentRecord.
                     1. Let _benv_ be the running execution context’s LexicalEnvironment.
                     1. Let _benvRec_ be _benv_'s EnvironmentRecord.
-                    1. Let _fobj_ be _benvRec_.GetBindingValue(_F_, *false*).
-                    1. Assert: _fobj_ is never an abrupt completion.
+                    1. Let _fobj_ be ! _benvRec_.GetBindingValue(_F_, *false*).
                     1. Let _status_ be ? _genvRec_.SetMutableBinding(_F_, _fobj_, *false*).
```
Comment (ljharb) [inline spec.html]: this `_status_` is never used, should it be `Perform ? _genvRec_.SetMutableBinding(_F_, _fobj_, *false*).`?

File: spec.html [hunk 6]
```diff
@@ -36631,8 +36625,7 @@
               1. If replacing the |FunctionDeclaration| _f_ with a |VariableStatement| that has _F_ as a |BindingIdentifier| would not produce any Early Errors for _func_ and _F_ is not an element of BoundNames of _argumentsList_, then
                 1. NOTE A var binding for _F_ is only instantiated here if it is neither a VarDeclaredName, the name of a formal parameter, or another |FunctionDeclaration|.
                 1. If _instantiatedVarNames_ does not contain _F_, then
-                  1. Let _status_ be _varEnvRec_.CreateMutableBinding(_F_).
-                  1. Assert: _status_ is never an abrupt completion.
+                  1. Let _status_ be ! _varEnvRec_.CreateMutableBinding(_F_).
```
Comment (ljharb) [inline spec.html]: this `_status_` is never used, should it be `Perform ! _varEnvRec_.CreateMutableBinding(_F_).`?

File: spec.html [hunk 7]
```diff
@@ -33617,11 +33611,11 @@
...
               1. Let _len_ be ? ToLength(? Get(_val_, `"length"`)).
               1. Repeat while _I_ &lt; _len_,
-                1. Let _newElement_ be ? InternalizeJSONProperty(_val_, ToString(_I_)).
+                1. Let _newElement_ be ? InternalizeJSONProperty(_val_, ! ToString(_I_)).
                 1. If _newElement_ is *undefined*, then
-                  1. Let _status_ be ? _val_.[[Delete]](ToString(_I_)).
+                  1. Let _status_ be ? _val_.[[Delete]](! ToString(_I_)).
                 1. Else,
-                  1. Let _status_ be ? CreateDataProperty(_val_, ToString(_I_), _newElement_).
+                  1. Let _status_ be ? CreateDataProperty(_val_, ! ToString(_I_), _newElement_).
```
Comment (ljharb) [inline spec.html]: unused `_status_`

File: spec.html [hunk 8]
```diff
@@ -31447,7 +31441,7 @@
           1. Let _A_ be ? AllocateTypedArray(_C_, _captured_).
           1. Let _n_ be 0.
           1. For each element _e_ of _kept_
-            1. Let _status_ be ? Set(_A_, ToString(_n_), _e_, *true* ).
+            1. Let _status_ be ? Set(_A_, ! ToString(_n_), _e_, *true* ).
```
Comment (ljharb) [inline spec.html]: unused `_status_`

File: spec.html [hunk 9]
```diff
@@ -30498,11 +30492,11 @@
           1. Let _A_ be ? ArraySpeciesCreate(_O_, _actualDeleteCount_).
           1. Let _k_ be 0.
           1. Repeat, while _k_ &lt; _actualDeleteCount_
-            1. Let _from_ be ToString(_actualStart_+_k_).
+            1. Let _from_ be ! ToString(_actualStart_+_k_).
             1. Let _fromPresent_ be ? HasProperty(_O_, _from_).
             1. If _fromPresent_ is *true*, then
               1. Let _fromValue_ be ? Get(_O_, _from_).
-              1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ToString(_k_), _fromValue_).
+              1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ! ToString(_k_), _fromValue_).
```
Comment (ljharb) [inline spec.html]: unused `_status_`

File: spec.html [hunk 10]
```diff
@@ -30289,11 +30283,11 @@
           1. Let _A_ be ? ArraySpeciesCreate(_O_, _count_).
           1. Let _n_ be 0.
           1. Repeat, while _k_ &lt; _final_
-            1. Let _Pk_ be ToString(_k_).
+            1. Let _Pk_ be ! ToString(_k_).
             1. Let _kPresent_ be ? HasProperty(_O_, _Pk_).
             1. If _kPresent_ is *true*, then
               1. Let _kValue_ be ? Get(_O_, _Pk_).
-              1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ToString(_n_), _kValue_ ).
+              1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ! ToString(_n_), _kValue_ ).
```
Comment (ljharb) [inline spec.html]: unused `_status_`

File: spec.html [hunk 11]
```diff
@@ -29905,7 +29899,7 @@
             1. Let _k_ be _len_ + _n_.
             1. If _k_ < 0, let _k_ be 0.
           1. Repeat, while _k_ < _len_
-            1. Let _elementK_ be the result of ? Get(_O_, ToString(_k_)).
+            1. Let _elementK_ be the result of ? Get(_O_, ! ToString(_k_)).
             1. If SameValueZero(_searchElement_, _elementK_) is *true*, return *true*.
```
Comment (ljharb) [inline spec.html]: should `SameValueZero` also get a `!` prefix throughout?

File: spec.html [hunk 12]
```diff
@@ -29772,13 +29766,13 @@
...
           1. Repeat, while _k_ &lt; _len_
-            1. Let _Pk_ be ToString(_k_).
+            1. Let _Pk_ be ! ToString(_k_).
             1. Let _kPresent_ be ? HasProperty(_O_, _Pk_).
             1. If _kPresent_ is *true*, then
               1. Let _kValue_ be ? Get(_O_, _Pk_).
               1. Let _selected_ be ToBoolean(? Call(_callbackfn_, _T_, &laquo; _kValue_, _k_, _O_ &raquo;)).
               1. If _selected_ is *true*, then
-                1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ToString(_to_), _kValue_).
+                1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ! ToString(_to_), _kValue_).
```
Comment (ljharb) [inline spec.html]: unused `_status_`

File: spec.html [hunk 13]
```diff
@@ -29597,16 +29591,16 @@
...
                 1. If _exists_ is *true*, then
                   1. Let _subElement_ be ? Get(_E_, _P_).
-                  1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ToString(_n_), _subElement_).
+                  1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ! ToString(_n_), _subElement_).
                 1. Increase _n_ by 1.
                 1. Increase _k_ by 1.
             1. Else _E_ is added as a single item rather than spread,
               1. If _n_&ge;2<sup>53</sup>-1, throw a *TypeError* exception.
-              1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ToString(_n_), _E_).
+              1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ! ToString(_n_), _E_).
```
Comment (ljharb) [inline spec.html]: unused `_status_`

File: spec.html [hunk 14]
```diff
@@ -29597,16 +29591,16 @@
               1. Let _len_ be ? ToLength(? Get(_E_, `"length"`)).
               1. If _n_ + _len_ &gt; 2<sup>53</sup>-1, throw a *TypeError* exception.
               1. Repeat, while _k_ &lt; _len_
-                1. Let _P_ be ToString(_k_).
+                1. Let _P_ be ! ToString(_k_).
                 1. Let _exists_ be ? HasProperty(_E_, _P_).
                 1. If _exists_ is *true*, then
                   1. Let _subElement_ be ? Get(_E_, _P_).
-                  1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ToString(_n_), _subElement_).
+                  1. Let _status_ be ? CreateDataPropertyOrThrow(_A_, ! ToString(_n_), _subElement_).
```
Comment (ljharb) [inline spec.html]: unused `_status_`

File: spec.html [hunk 15]
```diff
@@ -29275,15 +29269,15 @@
                 1. Let _numberOfCaptures_ be max(_numberOfCaptures_-1, 0).
                 1. Let _i_ be 1.
                 1. Repeat, while _i_ &le; _numberOfCaptures_.
-                  1. Let _nextCapture_ be ? Get(_z_, ToString(_i_)).
-                  1. Perform CreateDataProperty(_A_, ToString(_lengthA_), _nextCapture_).
+                  1. Let _nextCapture_ be ? Get(_z_, ! ToString(_i_)).
+                  1. Perform CreateDataProperty(_A_, ! ToString(_lengthA_), _nextCapture_).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 16]
```diff
@@ -29275,15 +29269,15 @@
...
                   1. Let _i_ be _i_ +1.
                   1. Let _lengthA_ be _lengthA_ +1.
                   1. If _lengthA_ = _lim_, return _A_.
                 1. Let _q_ be _p_.
           1. Let _T_ be a String value equal to the substring of _S_ consisting of the elements at indices _p_ (inclusive) through _size_ (exclusive).
           1. Assert: The following call will never result in an abrupt completion.
-          1. Perform CreateDataProperty(_A_, ToString(_lengthA_), _T_ ).
+          1. Perform CreateDataProperty(_A_, ! ToString(_lengthA_), _T_ ).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 17]
```diff
@@ -29267,7 +29261,7 @@
               1. Else _e_ &ne; _p_,
                 1. Let _T_ be a String value equal to the substring of _S_ consisting of the elements at indices _p_ (inclusive) through _q_ (exclusive).
                 1. Assert: The following call will never result in an abrupt completion.
-                1. Perform CreateDataProperty(_A_, ToString(_lengthA_), _T_).
+                1. Perform CreateDataProperty(_A_, ! ToString(_lengthA_), _T_).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 18]
```diff
@@ -29089,7 +29083,7 @@
                 1. Else, return _A_.
               1. Else _result_ is not *null*,
                 1. Let _matchStr_ be ? ToString(? Get(_result_, `"0"`)).
-                1. Let _status_ be CreateDataProperty(_A_, ToString(_n_), _matchStr_).
+                1. Let _status_ be CreateDataProperty(_A_, ! ToString(_n_), _matchStr_).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 19]
```diff
@@ -28991,7 +28985,7 @@
               1. Else, _fullUnicode_ is *false*,
                 1. Assert: _captureI_ is a List of code units.
                 1. Let _capturedValue_ be a string consisting of the code units of _captureI_.
-              1. Perform CreateDataProperty(_A_, ToString(_i_) , _capturedValue_).
+              1. Perform CreateDataProperty(_A_, ! ToString(_i_) , _capturedValue_).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 20]
```diff
@@ -27001,14 +26995,14 @@
...
+                1. Perform CreateDataProperty(_A_, ! ToString(_lengthA_), _T_).
                 1. Assert: The above call will never result in an abrupt completion.
                 1. Increment _lengthA_ by 1.
                 1. If _lengthA_ = _lim_, return _A_.
                 1. Let _p_ be _e_.
                 1. Let _q_ be _p_.
           1. Let _T_ be a String value equal to the substring of _S_ consisting of the code units at indices _p_ (inclusive) through _s_ (exclusive).
-          1. Perform CreateDataProperty(_A_, ToString(_lengthA_), _T_).
+          1. Perform CreateDataProperty(_A_, ! ToString(_lengthA_), _T_).
           1. Assert: The above call will never result in an abrupt completion.
```
Comment (ljharb) [inline spec.html]: `!` prefix on the above `CreateDataProperty`, and remove this line?

File: spec.html [hunk 21]
```diff
@@ -27001,14 +26995,14 @@
               1. If _e_ = _p_, let _q_ be _q_+1.
               1. Else _e_ &ne; _p_,
                 1. Let _T_ be a String value equal to the substring of _S_ consisting of the code units at indices _p_ (inclusive) through _q_ (exclusive).
-                1. Perform CreateDataProperty(_A_, ToString(_lengthA_), _T_).
+                1. Perform CreateDataProperty(_A_, ! ToString(_lengthA_), _T_).
                 1. Assert: The above call will never result in an abrupt completion.
```
Comment (ljharb) [inline spec.html]: `!` prefix on the above `CreateDataProperty`, and remove this line?

File: spec.html [hunk 22]
```diff
@@ -16073,8 +16068,7 @@
             1. Let _TDZ_ be NewDeclarativeEnvironment(_oldEnv_).
             1. Let _TDZEnvRec_ be _TDZ_'s EnvironmentRecord.
             1. For each string _name_ in _TDZnames_, do
-              1. Let _status_ be _TDZEnvRec_.CreateMutableBinding(_name_, *false*).
-              1. Assert: _status_ is never an abrupt completion.
+              1. Let _status_ be ! _TDZEnvRec_.CreateMutableBinding(_name_, *false*).
```
Comment (ljharb) [inline spec.html]: unused `_status_`?

File: spec.html [hunk 23]
```diff
@@ -15784,8 +15780,7 @@
             1. Let _thisIterationEnv_ be NewDeclarativeEnvironment(_outer_).
             1. Let _thisIterationEnvRec_ be _thisIterationEnv_'s EnvironmentRecord.
             1. For each element _bn_ of _perIterationBindings_ do,
-              1. Let _status_ be _thisIterationEnvRec_.CreateMutableBinding(_bn_, *false*).
-              1. Assert: _status_ is never an abrupt completion.
+              1. Let _status_ be ! _thisIterationEnvRec_.CreateMutableBinding(_bn_, *false*).
```
Comment (ljharb) [inline spec.html]: unused `status`?

File: spec.html [hunk 24]
```diff
@@ -15158,7 +15154,7 @@
             1. Let _nextValue_ be IteratorValue(_next_).
             1. If _nextValue_ is an abrupt completion, set _iteratorRecord_.[[done]] to *true*.
             1. ReturnIfAbrupt(_nextValue_).
-            1. Let _status_ be CreateDataProperty(_A_, ToString(_n_), _nextValue_).
+            1. Let _status_ be CreateDataProperty(_A_, ! ToString(_n_), _nextValue_).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 25]
```diff
@@ -14498,10 +14495,9 @@
         1. For each element _d_ in _declarations_ do
           1. For each element _dn_ of the BoundNames of _d_ do
             1. If IsConstantDeclaration of _d_ is *true*, then
-              1. Let _status_ be _envRec_.CreateImmutableBinding(_dn_, *true*).
+              1. Let _status_ be ! _envRec_.CreateImmutableBinding(_dn_, *true*).
             1. Else,
-              1. Let _status_ be _envRec_.CreateMutableBinding(_dn_, *false*).
-            1. Assert: _status_ is never an abrupt completion.
+              1. Let _status_ be ! _envRec_.CreateMutableBinding(_dn_, *false*).
```
Comment (ljharb) [inline spec.html]: unused `status`?

File: spec.html [hunk 26]
```diff
@@ -14498,10 +14495,9 @@
         1. For each element _d_ in _declarations_ do
           1. For each element _dn_ of the BoundNames of _d_ do
             1. If IsConstantDeclaration of _d_ is *true*, then
-              1. Let _status_ be _envRec_.CreateImmutableBinding(_dn_, *true*).
+              1. Let _status_ be ! _envRec_.CreateImmutableBinding(_dn_, *true*).
```
Comment (ljharb) [inline spec.html]: unused `status`?

File: spec.html [hunk 27]
```diff
@@ -7757,7 +7754,7 @@
               1. If _index_ &lt; _len_, then
                 1. Let _g_ be MakeArgGetter(_name_, _env_).
                 1. Let _p_ be MakeArgSetter(_name_, _env_).
-                1. Perform _map_.[[DefineOwnProperty]](ToString(_index_), PropertyDescriptor{[[Set]]: _p_, [[Get]]: _g_, [[Enumerable]]: *false*, [[Configurable]]: *true*}).
+                1. Perform _map_.[[DefineOwnProperty]](! ToString(_index_), PropertyDescriptor{[[Set]]: _p_, [[Get]]: _g_, [[Enumerable]]: *false*, [[Configurable]]: *true*}).
```
Comment (ljharb) [inline spec.html]: prefix on `[[DefineOwnProperty]]`?

File: spec.html [hunk 28]
```diff
@@ -7745,7 +7742,7 @@
           1. Let _index_ be 0.
           1. Repeat while _index_ &lt; _len_ ,
             1. Let _val_ be _argumentsList_[_index_].
-            1. Perform CreateDataProperty(_obj_, ToString(_index_), _val_).
+            1. Perform CreateDataProperty(_obj_, ! ToString(_index_), _val_).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 29]
```diff
@@ -7712,7 +7709,7 @@
           1. Let _index_ be 0.
           1. Repeat while _index_ &lt; _len_,
             1. Let _val_ be _argumentsList_[_index_].
-            1. Perform CreateDataProperty(_obj_, ToString(_index_), _val_).
+            1. Perform CreateDataProperty(_obj_, ! ToString(_index_), _val_).
```
Comment (ljharb) [inline spec.html]: prefix on `CreateDataProperty`?

File: spec.html [hunk 30]
```diff
@@ -7476,7 +7474,7 @@
           1. If _succeeded_ is *false*, return *false*.
           1. While _newLen_ &lt; _oldLen_ repeat,
             1. Set _oldLen_ to _oldLen_ - 1.
-            1. Let _deleteSucceeded_ be _A_.[[Delete]](ToString(_oldLen_)).
+            1. Let _deleteSucceeded_ be _A_.[[Delete]](! ToString(_oldLen_)).
             1. Assert: _deleteSucceeded_ is not an abrupt completion.
```
Comment (ljharb) [inline spec.html]: add prefix on `[[Delete]]`, and remove this line?

File: spec.html [hunk 31]
```diff
@@ -7171,15 +7171,13 @@
...
             1. Else,
-              1. Let _status_ be _lexEnvRec_.CreateMutableBinding(_dn_, *false*).
-          1. Assert: _status_ is never an abrupt completion.
+              1. Let _status_ be ! _lexEnvRec_.CreateMutableBinding(_dn_, *false*).
         1. For each parsed grammar phrase _f_ in _functionsToInitialize_, do
           1. Let _fn_ be the sole element of the BoundNames of _f_.
           1. Let _fo_ be the result of performing InstantiateFunctionObject for _f_ with argument _lexEnv_.
-          1. Let _status_ be _varEnvRec_.SetMutableBinding(_fn_, _fo_, *false*).
-          1. Assert: _status_ is never an abrupt completion.
+          1. Let _status_ be ! _varEnvRec_.SetMutableBinding(_fn_, _fo_, *false*).
```
Comment (ljharb) [inline spec.html]: unused `_status_`?

File: spec.html [hunk 32]
```diff
@@ -7171,15 +7171,13 @@
           1. NOTE A lexically declared name cannot be the same as a function/generator declaration, formal parameter, or a var name. Lexically declared names are only instantiated here but not initialized.
           1. For each element _dn_ of the BoundNames of _d_ do
             1. If IsConstantDeclaration of _d_ is *true*, then
-              1. Let _status_ be _lexEnvRec_.CreateImmutableBinding(_dn_, *true*).
+              1. Let _status_ be ! _lexEnvRec_.CreateImmutableBinding(_dn_, *true*).
             1. Else,
-              1. Let _status_ be _lexEnvRec_.CreateMutableBinding(_dn_, *false*).
-          1. Assert: _status_ is never an abrupt completion.
+              1. Let _status_ be ! _lexEnvRec_.CreateMutableBinding(_dn_, *false*).
```
Comment (ljharb) [inline spec.html]: unused `status`?

File: spec.html [hunk 33]
```diff
@@ -7171,15 +7171,13 @@
           1. NOTE A lexically declared name cannot be the same as a function/generator declaration, formal parameter, or a var name. Lexically declared names are only instantiated here but not initialized.
           1. For each element _dn_ of the BoundNames of _d_ do
             1. If IsConstantDeclaration of _d_ is *true*, then
-              1. Let _status_ be _lexEnvRec_.CreateImmutableBinding(_dn_, *true*).
+              1. Let _status_ be ! _lexEnvRec_.CreateImmutableBinding(_dn_, *true*).
```
Comment (ljharb) [inline spec.html]: unused `status`?

File: spec.html [hunk 34]
```diff
@@ -7151,12 +7153,10 @@
           1. For each _n_ in _varNames_, do
             1. If _n_ is not an element of _instantiatedVarNames_, then
               1. Append _n_ to _instantiatedVarNames_.
-              1. Let _status_ be _varEnvRec_.CreateMutableBinding(_n_).
-              1. Assert: _status_ is never an abrupt completion.
+              1. Let _status_ be ! _varEnvRec_.CreateMutableBinding(_n_).
```
Comment (ljharb) [inline spec.html]: unused `status`?

File: spec.html [hunk 35]
```diff
@@ -7137,8 +7140,7 @@
           1. For each _n_ in _varNames_, do
             1. If _n_ is not an element of _instantiatedVarNames_, then
               1. Append _n_ to _instantiatedVarNames_.
-              1. Let _status_ be _envRec_.CreateMutableBinding(_n_).
-              1. Assert: _status_ is never an abrupt completion.
+              1. Let _status_ be ! _envRec_.CreateMutableBinding(_n_).
```
Comment (ljharb) [inline spec.html]: unused `status`?

**Comment by ljharb** [comment]
k sorry, i'll stop - we should probably check all the `Let _status_` and `CreateDataProperty` usages and that will cover the majority of my comments :-)

**Comment by domenic** [comment]
This change is great, and @ljharb's thorough review is great too. We should merge. We can incrementally fix remaining issues later; I kind of expect @anba will have a way of doing so in an automated fashion...

**Comment by ljharb** [comment]
Agreed, I don't think any of my changes should block this change, just cleanups that might be missed later if not caught now :-) this is a huge improvement imo.

**Comment by bterlson** [comment]
This PR isn't trying to be exhaustive :) Landing this now, please send PRs for all the others as you come across them!

Committed as 166ab2b (rebased, fixed merge conflicts).

**Comment by anba** [comment]
> I kind of expect @anba will have a way of doing so in an automated fashion...

Fixing all remaining issues will be a tedious work. I think it's only fair the proponents of the `!` prefix get a chance to improve their commit statistics. Here's an incomplete (!) list what needs to be changed. :stuck_out_tongue_closed_eyes: 

- All calls to: IsAccessorDescriptor, IsDataDescriptor, IsGenericDescriptor, ObjectCreate, ToBoolean, IsCallable, IsConstructor, SameValue, SameValueZero, SameValueNonNumber, IsPropertyKey, CreateIterResultObject, CreateBuiltinFunction, CreateRealm, CreateIntrinsics, SetRealmGlobalObject, NewGlobalEnvironment, NewDeclarativeEnvironment, NewObjectEnvironment, NewFunctionEnvironment, NewModuleEnvironment, GetThisEnvironment, GetNewTarget, GetGlobalObject, EnqueueJob, AddRestrictedFunctionProperties, CanonicalNumericIndexString, IsInteger, CreateArrayFromList, CreateListIterator, FunctionAllocate, FunctionInitialize, FunctionCreate, GeneratorFunctionCreate, MakeConstructor, MakeClassConstructor, MakeMethod, SetFunctionName, StringCreate, CreateUnmappedArgumentsObject, CreateMappedArgumentsObject, IntegerIndexedObjectCreate, ModuleNamespaceCreate, ...

Reference type:
- All calls to: GetBase, GetReferencedName, IsStrictReference, HasPrimitiveBase, IsPropertyReference, IsUnresolvableReference, IsSuperReference, GetThisValue

Declarative Environment Record:
- All calls to: HasBinding, CreateMutableBinding, CreateImmutableBinding, InitializeBinding, DeleteBinding

Other operations:
- ToObject in Reference::GetValue and Reference::PutValue
- CreateDataProperty in FromPropertyDescriptor
- ToString in "ToString Applied to the Number Type"
- ToNumber in CanonicalNumericIndexString
- ArrayCreate in CreateArrayFromList
- CreateDataProperty in CreateArrayFromList
- CreateDataProperty in CreateIterResultObject
- CreateMethodProperty in CreateListIterator
- DefinePropertyOrThrow in MakeConstructor
- ...

**Pull request closed without merging**